src/Components/Endpoints/gen/ComponentEndpointsGenerator.cs

```diff
@@ -0,0 +1,478 @@
+        return ComponentModel.FromType((INamedTypeSymbol)component!, routeAttribute, renderModeAttribute);
+    }
+
+    private bool IsComponent((ISymbol? candidate, INamedTypeSymbol componentInterface) tuple)
+    {
+        return tuple.candidate is INamedTypeSymbol componentType &&
+            ComponentCollector.IsComponent(componentType, tuple.componentInterface);
+    }
+
+    private bool FilterAssemblies((IAssemblySymbol assembly, IAssemblySymbol componentsAssembly) context)
+    {
+        var (assembly, componentsAssembly) = context;
+        if (assembly.Name.StartsWith("System.", StringComparison.Ordinal) ||
+            assembly.Name.StartsWith("Microsoft.", StringComparison.Ordinal))
+        {
+            // Filter out system assemblies as well as our components assemblies.
+            return false;
+        }
+
+        if (assembly.Modules.Skip(1).Any())
+        {
+            return false;
+        }
+        var module = assembly.Modules.SingleOrDefault();
+        if (module == null)
+        {
+            return false;
+        }
+
+        foreach (var refIdentity in module.ReferencedAssemblies)
+        {
+            if (refIdentity == componentsAssembly.Identity)
+            {
+                return true;
+            }
+        }
+
+        return false;
+    }
+}
+
+public class ComponentCollector : SymbolVisitor
+{
+    public ComponentCollector(INamedTypeSymbol componentsInterface, INamedTypeSymbol routeAttribute, INamedTypeSymbol renderModeAttribute)
+    {
+        if (componentsInterface is null)
+        {
+            throw new ArgumentNullException(nameof(componentsInterface));
+        }
+
+        if (routeAttribute is null)
+        {
+            throw new ArgumentNullException(nameof(routeAttribute));
+        }
+
+        ComponentsInterface = componentsInterface;
+        RouteAttribute = routeAttribute;
+        RenderModeAttribute = renderModeAttribute;
+    }
+
+    public List<ComponentModel>? Components { get; set; }
+
+    public INamedTypeSymbol ComponentsInterface { get; set; }
+
+    public INamedTypeSymbol RouteAttribute { get; set; }
+
+    public INamedTypeSymbol RenderModeAttribute { get; set; }
+
+    public override void VisitNamespace(INamespaceSymbol symbol)
+    {
+        foreach (var member in symbol.GetMembers())
+        {
+            member.Accept(this);
+        }
+    }
+
+    public override void VisitNamedType(INamedTypeSymbol symbol)
+    {
+        if (IsComponent(symbol, ComponentsInterface))
+        {
+            Components ??= new();
+            Components.Add(ComponentModel.FromType(symbol, RouteAttribute, RenderModeAttribute));
+        }
+    }
+
+    internal static bool IsComponent(INamedTypeSymbol candidate, INamedTypeSymbol componentInterface)
+    {
+        if (candidate.TypeKind != TypeKind.Class ||
+            candidate.IsAbstract ||
+            candidate.IsAnonymousType ||
+            candidate.DeclaredAccessibility != Accessibility.Public ||
```

Comment:
The runtime doesn't require component classes to be public (though I know the Razor compiler does). If this is only being used to generate a list of page components with routes then it's fine as-is, but if this is meant to be an exhaustive list of all component types then we should be sure to include nested types and nonpublic types.

Reply:
Actually I realised that while we can refer to internal component types, we can't refer to nested private component types. So we have to avoid ever assuming we can generate an exhaustive list of component types. If this is only used for routable components then I guess that is fine.

Reply:
https://github.com/dotnet/razor/blob/69a87847a50299d3d98c5f70018afec9e49b4162/src/Compiler/Microsoft.CodeAnalysis.Razor/src/ComponentDetectionConventions.cs#L8-L46

The runtime doesn't enforce this constraint because the Razor compiler already does it. We had this discussion a few years ago with Ryan. We want to be in a position where you can define components in a way that is not discoverable or referenceable by the compiler and split the world between Razor syntax and non-Razor syntax.

We gain very little by supporting non-public/nested, etc. components, and we drastically constrain ourselves for the future.

Reply:
The fact remains that we do already support it. The situation with the Razor compiler is that if you want your component to show up publicly you have to make it public, but for internal use it's just C# classes and you are free to use any C# language features.

Reply:
The part I remember was specifically about *parameter* visibility, and even then, was about framework-supplied components not application-supplied components: https://github.com/dotnet/aspnetcore/issues/37840#issuecomment-954602593 Regardless, our conversations from 3 years ago are less important than what actually works in practice and hence what counts as backcompat.

> Even if we think this is not the case, I am ok for any code gen feature like this not working on those scenarios. The outcome is that either your page does not get automatically discovered or the render mode for your internal/private nested component is not taken into account.

That is OK. We just have to remember not to design anything where it becomes problematic for people to do this.

> That said, I think we are unnecessarily boxing ourselves here. There is little/none value from having a component being internal/private on apps and there are big drawbacks with regards to code generation.

I prefer to build a framework that naturally flows as part of the language. Components are like any other class. If there is value in making any class internal/private, the same rationale applies for components. It's very reasonable for an RCL to use components as internal implementation details without wanting them to be public.

Reply:
Keeping things open to pure C#-only components has been one of the strongest architectural moves we've pulled throughout Blazor and is partly what has stopped it from devolving into masses of custom template syntax as other UI frameworks and syntaxes have suffered in the past. It also keeps us open to major shifts in the future, such as entirely different code-based authoring models. So I am definitely keen not to compromise on that, even though it makes us think harder about framework design.

`.razor` is already a form of codegen, and doing further codegen on top of its conventions is perfectly fine and good. We just need to make sure we don't make non-`.razor` approaches impractical. We need to think of `.razor` as an extra layer within the system, not as the core.

For the specific features in this PR, I think the limitation of being public is fine. The reflection-based discovery within `Router` is already limited to public types so nothing weird is going to happen.